xarray/core/common.py

```diff
@@ -590,8 +602,11 @@ def rolling(self, dim=None, min_periods=None, center=False, **dim_kwargs):
         return self._rolling_cls(self, dim, min_periods=min_periods,
                                  center=center)
 
-    def coarsen(self, dim=None, boundary='exact', side='left',
-                coord_func='mean', **dim_kwargs):
+    def coarsen(self, dim: Optional[Mapping[Hashable, int]] = None,
+                boundary: str = 'exact',
+                side: Union[str, Mapping[Hashable, str]] = 'left',
+                coord_func: str = 'mean',
```

Comment:
didn't you say that all dims should be Hashable?

Reply:
I think `Mapping[Any, str]` and `Mapping[Hashable, str]` are equivalent -- every mapping key must be hashable.

Reply:
No, there's absolutely nothing in Mapping saying that the keys must be hashable.

This is perfectly legit, and pretty useful too:
```
class AnyKeyDict(MutableMapping[K, V]):
    def __init__(self, *args: Tuple[K, V], **kwargs: V)::
       self.mapping = {}
       for k, v in args:
           self[k] = v
       for k, v in kwargs.items():
           self[k] = v

    def __setitem__(self, k: K, v: V) -> None:
        try:
            self.mapping[k] = v
        except TypeError:
            self.mapping[pickle.dumps(k)] = v

    ...
```
This can be useful in a few borderline scenarios; for example when de-serialising msgpack data from a non-Python producer, as in msgpack map keys can be anything. Also I'd love to see something like this:
```
class DataArray:
    def __setitem__(self, k, v):
        if isinstance(k, Mapping):
            view = self.sel(**k)
        else:
            view = self.isel(**k)
        view[:] = v
```

Reply:
OK, good point.

It turns out we actually do already support this for `DataArray` :)
```
>>> data = xarray.DataArray([0, 0, 0], dims='x')
>>> data[{'x': 1}] = 1
>>> data
<xarray.DataArray (x: 3)>
array([0, 1, 0])
Dimensions without coordinates: x
```